### PR TITLE
fields query cause panic.

### PR DESCRIPTION
**Pull request by shimastripe**
## WHY

- fields query cause a panic and unpleasant behaviors.

Current status is as follows.

1. When the nest of the source is not a structure, panic occurs.

    ```
    ?fields=id.id
    ```

1. When specify a nonexistent field, it is ignored.

    ```
    ?fields=id,hogehoge
    ```

    ```json
    [
        {
            "id": 1
        }
    ]
    ```

1. It should return error if nest query is specified in spite of the structure is null.

    ```go
    User{
        AccountName: null,
    }
    ```

    ```
    ?fields=account_name.id
    ```

    ```json
        [
            {
                "account_name": null
            }
        ]
    ```

## WHAT

1. Checked the type of argument.
2. Checked the specified field whether it exists in the structure.
3. Checked whether the structure and the specified field is null or not.

### CODE REVIEW

File: _example/controllers/company.go [hunk 1]
```diff
@@ -58,16 +58,23 @@ func GetCompanies(c *gin.Context) {
 		// 1.0.0 <= this version < 2.0.0 !!
 	}
 
-	fieldMap := []map[string]interface{}{}
+	fieldMapArr := []map[string]interface{}{}
```
Comment (dtan4) [inline _example/controllers/company.go]: `fieldMapArr` seems not so good. How about `fieldMaps`?

File: _example/helper/field.go [hunk 2]
```diff
@@ -165,16 +178,21 @@ func FieldToMap(model interface{}, fields map[string]interface{}) map[string]int
...
 			v := fields[jsonKey]
+			var err error
 
 			if vs.Field(i).Kind() == reflect.Ptr {
 				if !vs.Field(i).IsNil() {
 					if v == nil {
 						u[jsonKey] = vs.Field(i).Elem().Interface()
 					} else {
-						u[jsonKey] = FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
+						u[jsonKey], err = FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
```
Comment (dtan4) [inline _example/helper/field.go]: Return err at this point.

```var
u[jsonKey], err := FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
if err != nil {
// ...
}
```

File: _example/helper/field.go [hunk 3]
```diff
@@ -184,24 +202,40 @@ func FieldToMap(model interface{}, fields map[string]interface{}) map[string]int
 					if v == nil {
 						fieldMap = append(fieldMap, s.Index(i).Interface())
 					} else {
+						nestFieldMap := make(map[string]interface{})
+
 						if s.Index(i).Kind() == reflect.Ptr {
-							fieldMap = append(fieldMap, FieldToMap(s.Index(i).Elem().Interface(), v.(map[string]interface{})))
+							nestFieldMap, err = FieldToMap(s.Index(i).Elem().Interface(), v.(map[string]interface{}))
```
Comment (dtan4) [inline _example/helper/field.go]: ditto.

File: _example/helper/field.go [hunk 4]
```diff
@@ -184,24 +202,40 @@ func FieldToMap(model interface{}, fields map[string]interface{}) map[string]int
...
 						} else {
-							fieldMap = append(fieldMap, FieldToMap(s.Index(i).Interface(), v.(map[string]interface{})))
+							nestFieldMap, err = FieldToMap(s.Index(i).Interface(), v.(map[string]interface{}))
```
Comment (dtan4) [inline _example/helper/field.go]: ditto/

File: _example/helper/field.go [hunk 5]
```diff
@@ -184,24 +202,40 @@ func FieldToMap(model interface{}, fields map[string]interface{}) map[string]int
...
 				}
 
 				u[jsonKey] = fieldMap
 			} else {
 				if v == nil {
 					u[jsonKey] = vs.Field(i).Interface()
 				} else {
-					u[jsonKey] = FieldToMap(vs.Field(i).Interface(), v.(map[string]interface{}))
+					nestFieldMap := make(map[string]interface{})
+					nestFieldMap, err = FieldToMap(vs.Field(i).Interface(), v.(map[string]interface{}))
```
Comment (dtan4) [inline _example/helper/field.go]: ditto.

**Comment by shimastripe** [comment]
Since these are all through the 233 line, so I summarize err. But it is still hard to see. . .
I fix later.

**Comment by shimastripe** [comment]
@dtan4 I fix. Please check again!

File: _example/helper/field.go [hunk 6]
```diff
@@ -165,16 +178,25 @@ func FieldToMap(model interface{}, fields map[string]interface{}) map[string]int
 
 		if contains(fields, jsonKey) {
 			v := fields[jsonKey]
+			var err error
```
Comment (dtan4) [inline _example/helper/field.go]: Is this really needed?

**Comment by shimastripe** [comment]
@dtan4 The following code was warned:

```go
u[jsonKey], err := FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
// => Warning: Go fails to infer type in assignment: “non-name on left side of :=”
```

The left-hand-side of := must be a list of identifiers, but `u[jsonKey]` is not identifiers.
So I can't use the shorthand declaration.

File: _example/helper/field.go (hunk 6 above)
Comment (shimastripe) [inline _example/helper/field.go]: @dtan4 The following code was warned:

```go
u[jsonKey], err := FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
// => Warning: Go fails to infer type in assignment: “non-name on left side of :=”
```

The left-hand-side of := must be a list of identifiers, but `u[jsonKey]` is not identifiers.
So I can't use the shorthand declaration.

Comment (dtan4) [inline _example/helper/field.go]: okay, how about this:

```go
k, err := FieldToMap(vs.Field(i).Elem().Interface(), v.(map[string]interface{}))
if err != nil {
// ...
}

u[jsonKey] = k
```

Comment (shimastripe) [inline _example/helper/field.go]: Thanks for your review! I fixed.

Since the mistake was found, it was fixed also.

**Comment by dtan4** [comment]
LGTM 👍

**Pull request merged**